Adds function which calculates sites selection frequency for a set of provided samples

Rosejoycrocker opened the pull request:
- allows plotting of site selection frequency choropleth maps using `create_map!` to plot frequencies.
- can take output of `find_robust()` to plot site selection frequency of robust scenarios.

ConnectedSystems commented:
Hi Rose, are there commits missing in this PR? I only see a function that collates frequency of seeding deployments - no `create_map()` or similar.

ConnectedSystems commented:
Marking this issue as contributing to resolving #254

Rosejoycrocker commented:
Hi Takuya, I realised you can use your create_map!() function with a suitable geojson and data = frequencies calculated from seeded_sites_frequency() to plot the frequncy maps (as far as I can see).

ConnectedSystems commented:
Oh I see, thanks!

ConnectedSystems commented on src/analysis/site_selection.jl:
```diff
@@ -0,0 +1,36 @@
+RCP60=[ ... frequency of selection for each site ...]]`
+
+"""
+function seeded_sites_frequency(rs::ResultSet, scens::NamedTuple)::NamedArray
```
On `scens` being a `namedtuple`, don't make any changes now but I am flagging that this is probably a roundabout way of handling multiple cases.

I've used this pattern before, but I think in cases like this where there's only a single "class" of data (i.e., data for each RCP), it's probably better to make things more flexible:

```julia
function seeded_sites_frequency(rs::ResultSet, rcps...)
    # ids will be each entry in `rcps`
end
```

Rosejoycrocker replied:
I made it a NamedTuple to work with the output of find_robust, which I imagine will be the most common use case of plotting robust scenario site selection frequencies

ConnectedSystems replied:
Yeah, I wasn't sure if that was the best way to do it  when I wrote that, was a bit rushed. Need to think a bit more about it.

Rosejoycrocker replied:
All good, should I change it to a variable number of inputs anyway?

ConnectedSystems replied:
No, hold off for now - I/we need to discuss what it should look like and how it might ease/complicate other analyses

Rosejoycrocker replied:
ok cool :)

Rosejoycrocker replied:
how's intervention_frequency() for the function name and scen_indices for the the scenario index NamedTuple?

Rosejoycrocker replied:
I could add a 'log' input which indicates shade of seed log so it does both.

ConnectedSystems replied:
I like `intervention_frequency()`

```julia

function intervention_frequency(rs, scen_indices, log_type)
    # This is just an example, really I'd pull out ResultSet fieldnames that end in "_log"
    occursin(log_type, [:seed, :shade, :fog]) || DomainError("Unsupported log")
   
    log = getfield(rs, Symbol("$(log_type)_log")))

    # Extract data from `log`
    # ...
end
```

ConnectedSystems replied:
`DomainError` is probably not the right error type to throw either - AttributeError? ValueError? I'll leave it to you.

Rosejoycrocker replied:
cool, I'll add this, thanks

ConnectedSystems reviewed (changes requested):
Requesting some minor adjustments. Some relate to making the user experience a little smoother

ConnectedSystems commented on src/analysis/site_selection.jl:
```diff
@@ -0,0 +1,36 @@
+- `scens` : contains scenario ids for scenarios satisfying the condition of interest.
+
+# Returns
+NamedArray, where each entry relates to an RCP of interest, e.g., `[RCP45=[... frequency of selection for each site ...]; 
```
Just realised - wouldn't this be better with reefs/sites as rows and RCPs as columns?

Not sure if the depiction above is accurate now either, it makes it look like an array of paired arrays.

Rosejoycrocker commented on src/analysis/site_selection.jl:
```diff
@@ -0,0 +1,36 @@
+    # retrieve RCPs
+    rcps = keys(scens)
+    # create frequencies storage container
+    seeded_sites_store = NamedArray(zeros(length(rcps), size(rs.site_data, 1)))
```
I can't import n_locations?

ConnectedSystems replied:
It doesn't exist for ResultSets yet, but it should still pull in the one for `Domain`. Just checking: where are you importing from?

Rosejoycrocker replied:
oh right, I can write one for ResultSet

Rosejoycrocker replied:
I put the `n_locations(rs)` function in the ResultSet file, but should it be with the Domain one?

Rosejoycrocker commented:
Hey @ConnectedSystems, I think I've addressed the comments. I'm not sure I've used NamedDimsArray in the way you'd like for the output, so let me know if I should change it.

Rosejoycrocker commented:
Ok updated the doc string and added axis keys, let me know if there's anything else

ConnectedSystems commented on src/analysis/site_selection.jl:
```diff
@@ -0,0 +1,57 @@
+    n_locs = n_locations(rs)
+
+    # create frequencies storage container
+    seeded_sites_store = NamedDimsArray(KeyedArray(zeros(n_locs, length(rcps)), (1:n_locs, [rcps[k] for k in eachindex(rcps)])), (:locations, :rcps))
```
Rose, please use the simpler method to creating named axes:

```julia
NamedDimsArray(zeros(n_locs, length(rcps)), locations=1:n_locs, rcps=rcps)
```

I'm not sure if this handles some edge case, so please double check that it's needed.
`[rcps[k] for k in eachindex(rcps)]`

Rosejoycrocker replied:
it errored previously without the vector creation because rcps is a Tuple

ConnectedSystems replied:
Okay, but if it's a tuple of values isn't looping over each index a convoluted way of converting to a vector?

`[rcps[k] for k in eachindex(rcps)]`

vs

```
# Indicating to Julia what the type is makes a tiny bit faster
Symbol[rcps...]
```

ConnectedSystems replied:
Or you could just make `rcps` a vector on L 44, then no conversion necessary

Rosejoycrocker replied:
Just realised I can use `collect()`, I always forget you can use it on tuples of strings/symbols as well

ConnectedSystems commented on src/analysis/site_selection.jl:
```diff
@@ -0,0 +1,57 @@
+
+    for rcp in rcps
+        # select scenarios satisfying condition and sum up selection tally for each site
+        seed_log = dropdims(sum(interv_log[scenarios=scen_indices[rcp]], dims=2), dims=2)
```
Is there an issue with using the dimension names?

ConnectedSystems reviewed (changes requested):
Thanks Rose, one, possibly two, more changes

ConnectedSystems commented on src/analysis/site_selection.jl:
```diff
@@ -0,0 +1,56 @@
+robust = ADRIA.analysis.find_robust(rs, y, rule_func, [45, 60])
+
+# find site selection frequencies for all robust scenarios
+robust_selection_frequencies = intervention_frequency(rs, scen_indices)
```
Wondering how this is being used on your end? I hit an undefined error, and I can't see this file being included anywhere - doesn't seem to be part of the ADRIA package?

Rosejoycrocker replied:
sorry,I just realised I was supposed to ask you about this in the PR and forgot to- should the import of site_selection.jl be in the ADRIA.jl file or the analysis module or both?

Rosejoycrocker replied:
Thanks, so analysis files only need to be imported in the analysis module?

ConnectedSystems replied:
Yes - functions/methods should be grouped/compartmentalised by modules according to their purpose. Their files are grouped by folders where appropriate.

You should see this if you look at the package directory structure. You should be able to intuit what lives where, and how it all relates to each other. For example, all code for analysis live together, ditto for file handling, etc.

At the top level, we only need to import the sub-modules (and later, package extensions).
Then the user only needs to import ADRIA or the sub-modules as necessary.

Otherwise you get into a situation where it's a messy free-for-all.

ConnectedSystems commented:
Cleaned things up for you


```julia
using ADRIA, Statistics, Plots

rs = ADRIA.load_results("some result set")

tac = ADRIA.metrics.scenario_total_cover(rs)
rsv = ADRIA.metrics.scenario_rsv(rs)

# Create matrix of mean scenario outcomes
mean_tac = vec(mean(tac, dims=1))
mean_sv = vec(mean(rsv, dims=1))
y = hcat(mean_tac, mean_sv)

# Find all pareto optimal scenarios where all metrics >= 0.9
rule_func = x -> all(x .>= 0.9)
robust_scens = ADRIA.analysis.find_robust(rs, y, rule_func, [45, 60])

# Retrieve seeding intervention frequency for robust scenarios
robust_selection_frequencies = ADRIA.analysis.intervention_frequency(rs, robust_scens, :seed)

bar(robust_selection_frequencies, labels=["RCP45" "RCP60"], xlabel="Sites", ylabel="Times Selected")
```

![image](https://user-images.githubusercontent.com/4075136/220227797-0b3277f7-d494-422c-9c2a-62b241aa989f.png)
